contracts/System.sol

```diff
@@ -1,71 +1,85 @@
+import {IBurn} from "./interface/IBurn.sol";
+import {IValidatorSet} from "./interface/IValidatorSet.sol";
+import {ICandidateHub} from "./interface/ICandidateHub.sol";
+import {ILightClient} from "./interface/ILightClient.sol";
+import {ISystemReward} from "./interface/ISystemReward.sol";
+import {ISlashIndicator} from "./interface/ISlashIndicator.sol";
+import {IRelayerHub} from "./interface/IRelayerHub.sol";
+import {IPledgeAgent} from "./interface/IPledgeAgent.sol";
 
-  bool public alreadyInit;
 
+contract System is Registry {
+  
+  Registry private immutable s_registry;  
 
-  address public constant VALIDATOR_CONTRACT_ADDR = 0x0000000000000000000000000000000000001000;
-  address public constant SLASH_CONTRACT_ADDR = 0x0000000000000000000000000000000000001001;
-  address public constant SYSTEM_REWARD_ADDR = 0x0000000000000000000000000000000000001002;
-  address public constant LIGHT_CLIENT_ADDR = 0x0000000000000000000000000000000000001003;
-  address public constant RELAYER_HUB_ADDR = 0x0000000000000000000000000000000000001004;
-  address public constant CANDIDATE_HUB_ADDR = 0x0000000000000000000000000000000000001005;
-  address public constant GOV_HUB_ADDR = 0x0000000000000000000000000000000000001006;
-  address public constant PLEDGE_AGENT_ADDR = 0x0000000000000000000000000000000000001007;
-  address public constant BURN_ADDR = 0x0000000000000000000000000000000000001008;
-  address public constant FOUNDATION_ADDR = 0x0000000000000000000000000000000000001009;
+  AllContracts private s_allContracts;  
 
+  constructor(Registry registry) {
+    s_registry = registry;  
+  }
 
-  modifier onlyCoinbase() {
-  
-    require(msg.sender == block.coinbase, "the message sender must be the block producer");
+  function updateContractAddr( //placeholder
+    address valAddr,
+    address slashAddr,
+    address rewardAddr,
+    address lightAddr,
+    address relayerHubAddr,
+    address candidateHubAddr,
+    address govHubAddr,
+    address pledgeAgentAddr,
+    address burnAddr,
+    address foundationAddr
+  ) external pure {
+  }
   
+  modifier onlyCoinbase() {
+    require(msg.sender == _coinbase(), "the message sender must be the block producer");
     _;
   }
 
   modifier onlyZeroGasPrice() {
-    
-    require(tx.gasprice == 0 , "gasprice is not zero");
-    
+    require(_gasprice() == 0 , "gasprice is not zero");
     _;
   }
 
-  modifier onlyNotInit() {
-    require(!alreadyInit, "the contract already init");
+  modifier onlyNotInit() { //placeholder
+    //require(!s_alreadyInit, "the contract already init");
     _;
   }
 
-  modifier onlyInit() {
-    require(alreadyInit, "the contract not init yet");
+  modifier onlyInit() { //placeholder
+    //require(s_alreadyInit, "the contract not init yet");
     _;
   }
 
   modifier onlySlash() {
-    require(msg.sender == SLASH_CONTRACT_ADDR, "the msg sender must be slash contract");
+    require(msg.sender == address(safe_slashIndicator()), "the msg sender must be slash contract");
```

Comment:
As discussed let's drop the `_safe` prefix for system contracts, instead denoting non-platform contracts as `_unsafe`
```suggestion
    require(msg.sender == address(slashIndicator()), "the msg sender must be slash contract");
```